api/src/main/java/com/velocitypowered/api/command/Command.java

```diff
@@ -1,59 +1,73 @@
 package com.velocitypowered.api.command;
 
 import com.google.common.collect.ImmutableList;
+import com.velocitypowered.api.proxy.Player;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
-import org.checkerframework.checker.nullness.qual.NonNull;
 
 /**
- * Represents a command that can be executed by a {@link CommandSource}, such as a {@link
- * com.velocitypowered.api.proxy.Player} or the console.
+ * Represents a command that can be executed by a {@link CommandSource}, such as
+ * a {@link Player} or the console.
  */
-public interface Command {
+public interface Command<I extends CommandInvocation> {
 
   /**
-   * Executes the command for the specified {@link CommandSource}.
+   * Executes the command for the specified invocation.
    *
-   * @param source the source of this command
-   * @param args the arguments for this command
+   * @param invocation the invocation context
    */
-  void execute(CommandSource source, String @NonNull [] args);
```

Comment:
Still API-breaking here - is there a better solution?

Reply:
If a command class now doesn't implement `LegacyCommand`, `BrigadierCommand` or `RawCommand`, the fallback methods will be executed. On 2.0.0 we can open up the possibility for custom `CommandInvocation` factories and `Command` implementations. As it stands, I don't think there's a way to detect a class using the deprecated methods without calling them.